cmd/soroban-rpc/internal/db/transaction.go

```diff
@@ -0,0 +1,347 @@
+package db
+
+import (
+	"context"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"time"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/stellar/go/ingest"
+	"github.com/stellar/go/support/db"
+	"github.com/stellar/go/support/errors"
+	"github.com/stellar/go/support/log"
+	"github.com/stellar/go/xdr"
+
+	"github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/ledgerbucketwindow"
+)
+
+const (
+	transactionTableName = "transactions"
+)
+
+type Transaction struct {
+	Result           []byte   // XDR encoded xdr.TransactionResult
+	Meta             []byte   // XDR encoded xdr.TransactionMeta
+	Envelope         []byte   // XDR encoded xdr.TransactionEnvelope
+	Events           [][]byte // XDR encoded xdr.DiagnosticEvent
+	FeeBump          bool
+	ApplicationOrder int32
+	Successful       bool
+	Ledger           ledgerbucketwindow.LedgerInfo
+}
+
+// TransactionWriter is used during ingestion to write LCM.
+type TransactionWriter interface {
+	InsertTransactions(lcm xdr.LedgerCloseMeta) error
+	RegisterMetrics(ingest, insert, count prometheus.Observer)
+}
+
+// TransactionReader is used to serve requests and just returns a cloned,
+// read-only DB session to perform actual DB reads on.
+type TransactionReader interface {
+	NewTx(ctx context.Context) (TransactionReaderTx, error)
+}
+
+// TransactionReaderTx provides all of the public ways to read from the DB.
+// Note that `Done()` *MUST* be called to clean things up.
+type TransactionReaderTx interface {
+	GetTransaction(hash xdr.Hash) (Transaction, ledgerbucketwindow.LedgerRange, error)
+	GetLedgerRange() (ledgerbucketwindow.LedgerRange, error)
```

Comment:
Would it be better to have GetLedgerRange in db/ledger.go instead?

Reply:
maybe as an interface, yeah, but then TransactionReader depends on LedgerReader which would be a bunch of premature refactoring - I think it'll be better to change this once we flesh out the interfaces more with the other pending db work